src/state/util/calculateGuessResult.ts

```diff
@@ -0,0 +1,73 @@
+import { isEqual } from 'lodash';
+
+export type MastermindCharacterKey =
+  | 'a'
+  | 'b'
+  | 'c'
+  | 'd'
+  | 'e'
+  | 'f'
+  | 'g'
+  | 'h'
+  | 'i'
+  | 'j';
+
+export type GuessResultKey = 'x' | '-' | ' ';
+
+export type GuessResult = {
+  solved: boolean;
+  result: GuessResultKey[];
+};
+
+function calculateGuessResult(params: {
+  answer: MastermindCharacterKey[];
+  guess: MastermindCharacterKey[];
+}): GuessResult {
+  const { answer, guess } = params;
+
+  if (answer.length !== guess.length) {
+    throw new Error(
+      `params.guess.length = ${guess.length}, but answer.length = ${answer.length}`
+    );
+  }
+
+  if (isEqual(guess, answer)) {
+    return { solved: true, result: answer.map(() => 'x' as GuessResultKey) };
+  }
+
+  const answerTracker: string[] = answer.slice(0);
+  const guessTracker: string[] = guess.slice(0);
+
+  const result: GuessResultKey[] = [];
+
+  answer.forEach((value, index) => {
+    if (value === guess[index]) {
+      result.push('x');
+      answerTracker[index] = '';
+      guessTracker[index] = '';
+    }
+  });
+
+  answerTracker.forEach((value, index) => {
+    if (!value) return;
+    const guessIndex = guessTracker.findIndex(
+      (guessValue) => guessValue === value
+    );
+    if (guessIndex < 0) return;
+    result.push('-');
+    answerTracker[index] = '';
+    guessTracker[guessIndex] = '';
+  });
+
```

Comment:
If you saw what I said before, I was wrong - looks like it works. I may or may not have had some crucial lines commented out, you'll never know.

Do you think it would make sense to tackle the logic within one `forEach`? Although it's a bit basic of a solution, maybe something like below for the loop logic? I suppose it's not too much different, since I did  use an `includes`
```js
const caclulateResult = (params) => {
    const { guess, answer } = params;

    // create copy of guess array
    const guessTracker = guess.slice(0);

    // code clues to be returned to user
    const keyPegFeedback = [];

    // checking to see if guess matches answer
    answer.forEach((answerKey, index) => {
        if (answerKey === guessTracker[index]) {
            keyPegFeedback.push('x');
        } else if (guessTracker.includes(answerKey)) {
            keyPegFeedback.push('-');
            guessTracker[guessTracker.indexOf(answerKey)] += 'Checked';
        } else {
            keyPegFeedback.push(' ');
        }
    });

    return keyPegFeedback.sort((a, b) => b.localeCompare(a));
}

caclulateResult({guess: ['c', 'a', 'd', 'c', 'j'], answer: ['b', 'a', 'a', 'c', 'b']});
```